pinecone/client.go

```diff
@@ -1343,6 +1343,125 @@ func (i *InferenceService) Embed(ctx context.Context, in *EmbedRequest) (*infere
+//   - Query: (Required) The query to compare with documents.
+//   - Documents: (Required) A list of Document objects to be reranked.
+//   - RankFields: (Optional) A list of document fields to use for ranking.
+//   - ReturnDocuments: (Optional) Whether to include documents in the response. Defaults to true.
+//   - TopN: (Optional) How many documents to return. Defaults to the length of input Documents.
+//   - Parameters: (Optional) Additional model-specific parameters for the reranker
+//
+// [model]: https://docs.pinecone.io/guides/inference/understanding-inference#models
+type RerankRequest struct {
+	Model           string
+	Query           string
+	Documents       []Document
+	RankFields      *[]string
+	ReturnDocuments *bool
+	TopN            *int
+	Parameters      *map[string]string
+}
+
+// Represents a ranked document with a relevance score and an index position.
+//
+// Fields:
+//   - Document: The [Document].
+//   - Index: The index position of the document from the original request. This can be used
+//     to locate the position of the document relative to others described in the request.
+//   - Score: The relevance score of the document indicating how closely it matches the query.
+type RankedDocument struct {
+	Document *Document `json:"document,omitempty"`
+	Index    int       `json:"index"`
+	Score    float32   `json:"score"`
+}
+
+// RerankResponse is the result of a reranking operation.
+//
+// Fields:
+//   - Data: A list of documents which have been reranked. The documents are sorted in order of relevance,
+//     with the first being the most relevant.
+//   - Model: The model used to rerank documents.
+//   - Usage: Usage statistics for the reranking operation.
+type RerankResponse struct {
+	Data  []RankedDocument `json:"data,omitempty"`
+	Model string           `json:"model"`
+	Usage RerankUsage      `json:"usage"`
+}
+
+// Rerank documents with associated relevance scores that represent the relevance of each document
+// to the provided query using the specified model.
+//
+// Parameters:
+//   - ctx: A context.Context object controls the request's lifetime, allowing for the request
+//     to be canceled or to timeout according to the context's deadline.
+//   - in: A pointer to a [RerankRequest] object that contains the model, query, and documents to use for reranking.
+//
+// Example:
+//
+//	     ctx := context.Background()
+//
+//	     clientParams := pinecone.NewClientParams{
+//		        ApiKey:    "YOUR_API_KEY",
+//		        SourceTag: "your_source_identifier", // optional
+//	     }
+//
+//	     pc, err := pinecone.NewClient(clientParams)
+//	     if err != nil {
+//		        log.Fatalf("Failed to create Client: %v", err)
+//	     }
+//
+//	     rerankModel := "bge-reranker-v2-m3"
+//	     topN := 2
+//	     retunDocuments := true
+//	     documents := []pinecone.Document{
+//		        {"id": "vec1", "text": "Apple is a popular fruit known for its sweetness and crisp texture."},
```

Comment:
Nit: I actually wouldn't use the term `vec[whatever]` here. I think it's a little confusing, since what's in the "Document" objs we're passing in is _text_, not an embedding. Maybe just stick with "abc1" or whatever for `ID`?